### ISSUE TITLE
GPU slower than CPU and GPU-driver breaks down under heavy load

### ISSUE THREAD

**Issue by Endnutzer** [gpujs/gpu.js#360]
I have tested the Benchmark on http://gpu.rocks/ where i get the following results:
CPU: 3.222s
GPU: 0.111s --> faster

Now when i run the code attached i get the following results:
CPU: 4ms
GPU: 23ms --> slower

Did i do anything wrong in the code?

Furthermore if i make three for-loops the GPU-drivers breaks down. How to avoid that? Setting "loopMaxIterations: 2048" didn't help obviously.

[Example.zip](https://github.com/gpujs/gpu.js/files/2318720/Example.zip)

**Comment by robertleeplummerjr** [comment]
I think the overhead of using the GPU and the problem not being very computationally expensive may be partly to blame.  What hardware are you running?

**Comment by Endnutzer** [comment]
It was running on an i5 with integrated graphics.

**Comment by nickhingston** [comment]
Had a very brief look - think there is definitely a problem here.

On chrome on my 2018 Macbook pro - i'm seeing 17ms for CPU and ~600ms for GPU.

I would not necessarily expect the GPU to be faster than CPU for this particular use case, but it should not be 35x slower!  Seems readPixels (i.e. reading the response from GPU RAM) is taking over 500ms...I imagine we are trying to read way more data than we need to here.

**Comment by nickhingston** [comment]
actually - thinking about the problem a bit - you are using the GPU in a single 'thread' - there is no parallelism, the problem needs breaking up into discrete chunks to make full use of the pipeline  - and a GPU will definitely be quite a lot slower.  I'm still surprised readPixels is taking most of the time here, so will leave the problem open

**Comment by Endnutzer** [comment]
So there is room for improvements i would say :-) the project is very interesting, so i am looking forward to the next releases

**Comment by Endnutzer** [comment]
Do you have an example on how to break it inparts and compute parallel?

**Comment by nickhingston** [comment]
OK, had another look, there is no issue here.  GPU.js takes quite a lot of time to initialise ~100ms on my system.  In the process of my investigation I have found some performance issues, but I will raise these under a different ticket, as they are not related to this specifically.

If you were not to sum the result into a single value (i.e. have an array output to keep all the values in) then the vector processor can do a load of the Math.atan(i) * Math.tan(k) calculations in parallel.

I have edited your example to use the same calculation, but not added the values up in a loop.  Also some examples of how gpu.js & kernel init takes up time. 

On my system GPU is over 2x as fast and gets faster the larger you set resultWidth...
[example_array_output.zip](https://github.com/gpujs/gpu.js/files/2343108/example_array_output.zip)

Basically loops on the GPU are a bad idea.  It would be possible to add up the results in some kind of parallel fashion, but solution would be complex and probably best and to just do it on the CPU.
